core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java

```diff
@@ -0,0 +1,102 @@
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.file.meta;
+
+import alluxio.AlluxioURI;
+import alluxio.exception.ExceptionMessage;
+import alluxio.master.journal.JournalContext;
+import alluxio.proto.journal.File;
+import alluxio.proto.journal.Journal;
+import alluxio.retry.ExponentialBackoffRetry;
+import alluxio.retry.RetryPolicy;
+import alluxio.security.authorization.Mode;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.underfs.options.MkdirsOptions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * Utility methods for inodes.
+ */
+public final class InodeUtils {
+  private static final Logger LOG = LoggerFactory.getLogger(InodeUtils.class);
+  /** The base amount (exponential backoff) to sleep before retrying persisting an inode. */
+  private static final int PERSIST_WAIT_BASE_SLEEP_MS = 2;
+  /** Maximum amount (exponential backoff) to sleep before retrying persisting an inode. */
+  private static final int PERSIST_WAIT_MAX_SLEEP_MS = 1000;
+  /** The maximum retries for persisting an inode. */
+  private static final int PERSIST_WAIT_MAX_RETRIES = 50;
+
+  private InodeUtils() {} // prevent instantiation
+
+  /**
+   * Synchronously persists an {@link InodeDirectory} to the UFS. If concurrent calls are made, only
+   * one thread will persist to UFS, and the others will wait until it is persisted.
+   *
+   * @param dir the {@link InodeDirectory} to persist
+   * @param inodeTree the {@link InodeTree}
+   * @param mountTable the {@link MountTable}
+   * @param journalContext the journal context
+   * @throws IOException if the file fails to persist
+   */
+  public static void syncPersistDirectory(InodeDirectory dir, InodeTree inodeTree,
+      MountTable mountTable, JournalContext journalContext) throws IOException {
+    RetryPolicy retry =
+        new ExponentialBackoffRetry(PERSIST_WAIT_BASE_SLEEP_MS, PERSIST_WAIT_MAX_SLEEP_MS,
+            PERSIST_WAIT_MAX_RETRIES);
+    while (dir.getPersistenceState() != PersistenceState.PERSISTED) {
+      if (dir.compareAndSwap(PersistenceState.NOT_PERSISTED,
+          PersistenceState.TO_BE_PERSISTED)) {
+        boolean success = false;
+        StringBuilder errorDetails = new StringBuilder("Inode: ").append(dir.toString());
+        try {
+          AlluxioURI uri = inodeTree.getPath(dir);
+          MountTable.Resolution resolution = mountTable.resolve(uri);
+          String ufsUri = resolution.getUri().toString();
+          errorDetails.append(" ufsUri: ").append(ufsUri);
+          UnderFileSystem ufs = resolution.getUfs();
+          MkdirsOptions mkdirsOptions =
+              MkdirsOptions.defaults().setCreateParent(false).setOwner(dir.getOwner())
+                  .setGroup(dir.getGroup()).setMode(new Mode(dir.getMode()));
+          if (!ufs.isDirectory(ufsUri)) {
```

Comment:
This was what the previous "persist" codepath did. If the dir already exists, it will skip the `mkdirs`.

This will change if we update the UFS api.